Add support for LSP formatting `options` parameter

masad-frost opened the pull request:
This PR adds support for LSP formatting options https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#formattingOptions for the yapf

It adds explicit support for `tabSize` and `insertSpaces`, the rest of the options are not configurable in yapf. I did however add support for arbitrary configurations as specified in `[key: string]: boolean | integer | string;`.

I think this PR might break plugins as the API now passes `options` to formatter hooks, but not sure.

Nits are welcome as I don't normally read/write python.

masad-frost committed:
Update hookspec

masad-frost committed:
Add options support to yapf

masad-frost committed:
oops

ccordoba12 commented on pylsp/hookspecs.py:
```diff
@@ -80,12 +80,12 @@ def pylsp_folding_range(config, workspace, document):
 
 @hookspec(firstresult=True)
-def pylsp_format_document(config, workspace, document):
+def pylsp_format_document(config, workspace, document, options):
```
Unfortunately, this is not the right way of passing options to plugins. Plugins read options from the `config` object passed as the first parameter in this call.

You can see how that's handled by (for instance) the pycodestyle plugin here:

https://github.com/python-lsp/python-lsp-server/blob/develop/pylsp/plugins/pycodestyle_lint.py

So, what you need to do is basically something similar for Yapf.

masad-frost replied:
These are options passed with the request, not configuration options set ahead of time. Unless I'm misunderstanding

ccordoba12 replied:
Right, but all our plugins work like this. It makes little sense to expect that every request has its own set of configuration options because users can't control that on their editors/IDEs.

Instead, clients are expected to send a `workspace/didChangeConfiguration` request with a new set of options, those are saved in our config system, and then the next formatting or linting request will read and use them.

masad-frost replied:
The `textDocument/formatting` request has explicit formatting options, the configuration isn't something that is specified in the protocol, it's specified as `LSPAny`. There isn't a well-known way to specify formatting options in the workspace configuration. To add, not all the files follow the same formatting options so this isn't really a workspace-wide config.

> all our plugins work like this

Which plugins/requests are you referring to exactly? i.e. something that has a parameter that is instead replaced with a workspace configuration

> It makes little sense to expect that every request has its own set of configuration options because users can't control that on their editors/IDEs.

How so? Client implementations send formatting options based on user's options.

krassowski replied:
I think that @ccordoba12 meant workspace-wide settings as in:

https://github.com/python-lsp/python-lsp-server/blob/f19b3b9d44f4a2d5ab1e241f83a3fbc3390d25cc/pylsp/plugins/jedi_completion.py#L38-L47

But I agree with you @masad-frost here. For `textDocument/formatting` request the LSP protocol defines `DocumentFormattingParams`  which includes `options: FormattingOptions;`.

How IDEs decide to pass this options is irrelevant - they may hard-code the options, or expose them to user per-workspace or per-document - their choice.

ccordoba12 replied:
> For textDocument/formatting request the LSP protocol defines DocumentFormattingParams which includes options: FormattingOptions

You're right. I didn't know the LSP allows to pass options to this specific request.

masad-frost replied:
Oh, thanks @krassowski for clarifying the misunderstanding, maybe my PR description wasn't clear.

I'm currently traveling but can probably fix up the PR's tests in the coming week. Let me know if there's other comments. 🙏

ccordoba12 reviewed (commented):
Hey @masad-frost, thanks for your contribution! Please read my comment below to understand how to reorient your work to fit the design of the other plugins in this project.

krassowski commented:
The tests are failing on CI:

```
>           style_config = style.CreateStyleFromConfig(style_config)
E           NameError: name 'style' is not defined
```

masad-frost committed:
Add missing import

masad-frost commented:
Thanks for waiting! Got a little busy getting married :D

I rebased the branch and fixed the error in the tests https://github.com/python-lsp/python-lsp-server/pull/134/commits/22db74e6ec9b3a47c1a866f0e8a65bd0735a9a08

masad-frost committed:
Pass options through for format range

masad-frost commented:
Lint caught an issue where I wasn't passing `options` for range formatting!

ccordoba12 commented:
Hey @masad-frost, I fixed the linting errors reported here in another PR of mine.

Please let me know if you need help to fix the merge conflicts so we can merge this one.

masad-frost committed:
Merge remote-tracking branch 'origin/develop' into fm-yapf-formatting-options

masad-frost commented:
Done

masad-frost committed:
Update test_yapf_format.py

masad-frost commented:
Hmm, not sure why lint passes locally and crops up here only

masad-frost committed:
More lint

masad-frost commented:
Hmm, not sure why lint passed locally. Hopefully this is the last of it

ccordoba12 commented on pylsp/plugins/autopep8_format.py:
```diff
@@ -13,13 +13,13 @@
 
 @hookimpl(tryfirst=True)  # Prefer autopep8 over YAPF
-def pylsp_format_document(config, document):
+def pylsp_format_document(config, document, _options=None):
```
```suggestion
def pylsp_format_document(config, document, options=None):
```

Let's leave this parameter without an underscore, even if it's unused.

ccordoba12 commented on pylsp/plugins/autopep8_format.py:
```diff
@@ -13,13 +13,13 @@
 
 @hookimpl(tryfirst=True)  # Prefer autopep8 over YAPF
-def pylsp_format_range(config, document, range):  # pylint: disable=redefined-builtin
+def pylsp_format_range(config, document, range, _options=None):  # pylint: disable=redefined-builtin
```
```suggestion
def pylsp_format_range(config, document, range, options=None):  # pylint: disable=redefined-builtin
```

The same here.

ccordoba12 commented on test/plugins/test_yapf_format.py:
```diff
@@ -68,3 +68,30 @@ def test_line_endings(workspace, newline):
 
     assert res[0]['newText'] == f'import os{newline}import sys{2 * newline}dict(a=1){newline}'
+
+FOUR_SPACE_DOC = """def hello():
+    pass
+"""
```
Please move this variable to be after the import's section in this file because it's a constant.

ccordoba12 reviewed (commented):
Hey @masad-frost, last review then this should be ready.

I've also seen that you have struggled a lot with our linting tests, so I'll give you a hand with them if you're unable to fix them after this review.

masad-frost committed:
Address feedback

ccordoba12 committed:
Fix last linting issues

ccordoba12 committed:
Fix linting issues in tests

ccordoba12 reviewed (approved):
Thanks for your contribution @masad-frost!

masad-frost commented:
Hey @ccordoba12, when do you plan on tagging a new release? Wondering because we wanna upgrade, also maybe i can slip https://github.com/python-lsp/python-lsp-server/pull/136 before then.

ccordoba12 commented:
Perhaps in a month or so.

masad-frost commented:
Good time to bump to 1.5?

ccordoba12 commented:
Yep, I'll release it next week.
